plugins/newsdownloader.koplugin/epubdownloadbackend.lua

```diff
@@ -24,6 +24,54 @@ local max_redirects = 5; --prevent infinite redirects
 local TIMEOUT_CODE = "timeout" -- from socket.lua
 local MAXTIME_CODE = "maxtime reached" -- from sink_table_with_maxtime
 
+-- filter HTML using CSS selector
+local function filter(text, element)
+    local htmlparser = require("htmlparser")
+    local root = htmlparser.parse(text, 10000)
+    local filtered = nil
+    local selectors = {
+        "main",
+        "article",
+        "div#main",
+        "#main-article",
+        ".main-content",
+        "#body",
+        "#content",
+        ".content",
+        "div#article",
+        "div.article",
+        "div.post",
+        "div.post-outer",
+        ".l-root",
+        ".content-container",
+        ".StandardArticleBody_body",
+        "div#article-inner",
+        "div#newsstorytext",
+        "div.general",
+        }
+    if element then 
+        table.insert(selectors, 1, element)
+    end
+    for _, sel in ipairs(selectors) do
+       local elements = root:select(sel)
+       if elements then
+           for _, e in ipairs(elements) do
+               filtered = e:getcontent()
+               if filtered then 
+                   break
+               end
+           end
+           if filtered then
+              break 
+           end
+       end
+    end
+    if not filtered then 
+        return text 
+    end
+    return "<!DOCTYPE html><html><head></head><body>" .. filtered .. "</body></html>"
```

Comment:
Should this take the title from the original HTML document?

Reply:
I do not think such is needed as the title is already imported before the filter function is initiated and the title exists as epub title, so it is visible in the list.

Matter of preference nonetheless.

Reply:
This is supposed to go directly into an EPUB? Is this valid EPUB3? An EPUB content file would normally be expected to start with something like this since it's X(HT)ML:

```xml
<?xml version="1.0" encoding="utf-8"?>
<html xmlns="http://www.w3.org/1999/xhtml">
```

Reply:
Entire HTML in the default setting was going in anyway, and I've have had bad experience offering the later EPUB parser the HTML without this syntax header and footer. The rest of the program works the same after filtering so any changes done towards EPUB are there anyway. Instead of getting the full html, it just gets the things in the chosen selector and this header and footer.

Reply:
You mean that crengine doesn't use the XML parser then?

Reply:
Well, the rest of the program uses the old regex (kindly avoids summoning Zalgo) and then maybe xml parser, but the htmlparser is only used here.

Reply:
(crengine = our EPUB/HTML renderer)

With EPUB, crengine will always use the XML parser (while for standalone HTML files, it uses the loosy HTML parser that do tag autoclose and stuff).
But crengine does not care much about the DOCTYPE/XML/HTML header - no matter what it says, it won't change crengine's parsing behaviour.

I guess htmlparser, when spitting out the HTML from the selected element, will spit out a perfectly balanced serialized DOM, so crengine will probably be happier with that than with the original possibly messy downloaded HTML.

And indeed, as these HTML files ar just added to an EPUB, they don't need a TITLE - it would not be used by crengine anyway.

Reply:
Two minor things, crengine (or alternatively MuPDF) is what renders the thing.

And Lua patterns aren't regex.

Anyway, like I said it can stay here.

Reply:
I see now, I was mixed up, but basically everything is fine in your opinion regarding this?

Reply:
If there's no reason not to use `<?xml version="1.0" encoding="utf-8"?><html xmlns="http://www.w3.org/1999/xhtml">`, I'd rather see it in there as a good example for people learning by looking at the output. One thing I didn't fully realize a few years ago is that what seems like allowing a fairly innocent shortcut that we all know isn't strictly right, just not wanting to nitpick too much, often somehow comes back to bite you in the future.